### ISSUE TITLE
app-compat: Assertion failed when running `node-gnu` with `app-elfloader`

### ISSUE THREAD

**Issue by razvand** [unikraft/unikraft#860]
### Describe the bug

When using [`app-elfloader`](https://github.com/unikraft/app-elfloader) (via the `run_app.sh` script in [`run-app-elfloader`](https://github.com/unikraft/run-app-elfloader)) to run the [`node-gnu` static PIE build](https://github.com/unikraft/static-pie-apps/tree/master/node), an assertion appears.

### Steps to reproduce

You can use the [`./do.sh`](https://github.com/unikraft-upb/scripts/tree/main/make-based/app-elfloader) script in the [`unikraft-upb/scripts` repository](https://github.com/unikraft-upb/scripts):

```console
$ ./do.sh setup
$ ./do.sh build
$ ./do.sh run_built node_static_gnu
```

### Expected behavior

A Node server should start, waiting for connections.

### Which architectures were you using or does this bug affect?

x86_64

### Which platforms were you using or does this bug affect?

kvm

### Relevant log output

```shell
getegid(0x47a45f880, 0x47a45f8b0, ...) = 0x0
futex(va:0x474ad113c, FUTEX_WAKE|FUTEX_PRIVATE_FLAG, 0x7fffffff) = OK
futex(va:0x474ad1130, FUTEX_WAKE|FUTEX_PRIVATE_FLAG, 0x7fffffff) = OK
futex(va:0x474ad1128, FUTEX_WAKE|FUTEX_PRIVATE_FLAG, 0x7fffffff) = OK
futex(va:0x474ad0b84, FUTEX_WAKE|FUTEX_PRIVATE_FLAG, 0x7fffffff) = OK
futex(va:0x474ad1174, FUTEX_WAKE|FUTEX_PRIVATE_FLAG, 0x7fffffff) = OK
futex(va:0x474ad1270, FUTEX_WAKE|FUTEX_PRIVATE_FLAG, 0x7fffffff) = OK
futex(va:0x474ad0c48, FUTEX_WAKE|FUTEX_PRIVATE_FLAG, 0x7fffffff) = OK
[    0.640313] CRIT: [libkvmplat] <trace.c @   41> RIP: 0000000471b680bc CS: 0008
[    0.643050] CRIT: [libkvmplat] <trace.c @   42> RSP: 000000047a45f930 SS: 0010 EFLAGS: 00010246
[    0.646195] CRIT: [libkvmplat] <trace.c @   44> RAX: 0000000000000001 RBX: 0000000000000000 RCX: 0000000471b67f8e
[    0.649920] CRIT: [libkvmplat] <trace.c @   46> RDX: 000000007fffffff RSI: 0000000000000000 RDI: 0000000000000000
[    0.653686] CRIT: [libkvmplat] <trace.c @   48> RBP: 000000047a45f950 R08: 00000010000362d0 R09: 0000000000000023
[    0.658330] CRIT: [libkvmplat] <trace.c @   50> R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000005
[    0.662994] CRIT: [libkvmplat] <trace.c @   52> R13: 0000000474a0f680 R14: 0000000474a0f680 R15: 000000047a45fb30
[    0.666748] CRIT: [libkvmplat] <trace.c @   86> base is 0x47a45f950 caller is 0x471211cdb
[    0.669767] CRIT: [libkvmplat] <trace.c @   86> base is 0x47a45f990 caller is 0x4712599d2
[    0.672606] CRIT: [libkvmplat] <trace.c @   86> base is 0x47a45f9c0 caller is 0x471259eb7
[    0.675410] CRIT: [libkvmplat] <trace.c @   86> base is 0x47a45f9e0 caller is 0x47049f447
[    0.678215] CRIT: [libkvmplat] <trace.c @   86> base is 0x47a45fa00 caller is 0x4702ad157
[    0.681085] CRIT: [libkvmplat] <trace.c @   86> base is 0x47a45fbf0 caller is 0x4702ad696
[    0.683871] CRIT: [libkvmplat] <trace.c @   86> base is 0x47a45fc80 caller is 0x471b252fa
[    0.686907] CRIT: [libkvmplat] <trace.c @   86> base is 0x2 RIP: 0000000000104151 CS: 0008
[    0.689824] CRIT: [libkvmplat] <trace.c @   42> RSP: 00000000001ffee0 SS: 0010 EFLAGS: 00010002
[    0.694131] CRIT: [libkvmplat] <trace.c @   44> RAX: 0000000000000000 RBX: 0000000000000002 RCX: 0000000000000018
[    0.698077] CRIT: [libkvmplat] <trace.c @   46> RDX: 80000000001a2328 RSI: 0000000000000000 RDI: 00000000001a2329
[    0.701860] CRIT: [libkvmplat] <trace.c @   48> RBP: 00000000001ffef0 R08: 0000000000000046 R09: 00000000000000ff
[    0.705481] CRIT: [libkvmplat] <trace.c @   50> R10: 0000000000000002 R11: 0000000000000000 R12: 0000000000000000
[    0.709010] CRIT: [libkvmplat] <trace.c @   52> R13: 00000000001fff50 R14: 0000000474a0f680 R15: 000000047a45fb30
[    0.712652] CRIT: [libkvmplat] <trace.c @   86> base is 0x1ffef0 caller is 0x104f3a
[    0.715867] CRIT: [libkvmplat] <trace.c @   86> base is 0x1fff40 caller is 0x10ee3b
[    0.719147] CRIT: [libkvmplat] <trace.c @   86> base is 0x1ffef0 caller is 0x104f3a
```

**Comment by flpostolache** [comment]
I'll document what I've found till now, just to be written somewhere. I will also give the source names of some sources to easier to find the functions. The crash happens at the initialization of the openssl module, to be exact, in the `ossl_lib_ctx_get_data` function from `context.c`. The function attemps to read-lock the `ctx->lock` variable, which in our case is uninitialized(the structure is because it is a staically declared one). The initialization of this lock should take place in the `OPENSSL_init_crypto` function, when the `ossl_init_config_settings`(init.c) is called. This last function never gets called because before it, there is another `lock` on another rwlock that fails with `EDEADLK` and returns 0.

**Comment by flpostolache** [comment]
Update. The problem appears in `pthread_rwlock_wrlock` function. Here, the function checks if the writer thread tries to acquire the lock. In the case it does, it returns `EDEADLK`. In Unikraft's case what happens is that the lock structure was just initialised and the current writer member is set to 0. Because the TID is also set to 0, the call fails. The solution would be to change the TID to something other than 0.

**Comment by flpostolache** [comment]
@razvand This was solved by #885. I think we can close it.

**Comment by razvand** [comment]
Solved by #885 , as mentioned by @flpostolache
